// @aws-quickstart/eks-blueprints/dist/addons/opa-gatekeeper/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpaGatekeeperAddOn = void 0;
const helm_addon_1 = require("../helm-addon");
/**
 * Defaults options for the gatekeeper add-on
 */
const defaultProps = {
    name: 'gatekeeper',
    release: 'blueprints-addon-opa-gatekeeper',
    namespace: 'gatekeeper-system',
    chart: 'gatekeeper',
    repository: "https://open-policy-agent.github.io/gatekeeper/charts",
    version: '3.13.0'
};
class OpaGatekeeperAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(_clusterInfo) {
        return;
    }
    postDeploy(clusterInfo, _teams) {
        var _a;
        const chart = this.addHelmChart(clusterInfo, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        for (let provisioned of clusterInfo.getAllProvisionedAddons().values()) {
            chart.node.addDependency(provisioned);
        }
    }
}
exports.OpaGatekeeperAddOn = OpaGatekeeperAddOn;
