// @aws-sdk/client-eks/dist-cjs/EKS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EKS = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AssociateEncryptionConfigCommand_1 = require("./commands/AssociateEncryptionConfigCommand");
const AssociateIdentityProviderConfigCommand_1 = require("./commands/AssociateIdentityProviderConfigCommand");
const CreateAddonCommand_1 = require("./commands/CreateAddonCommand");
const CreateClusterCommand_1 = require("./commands/CreateClusterCommand");
const CreateFargateProfileCommand_1 = require("./commands/CreateFargateProfileCommand");
const CreateNodegroupCommand_1 = require("./commands/CreateNodegroupCommand");
const DeleteAddonCommand_1 = require("./commands/DeleteAddonCommand");
const DeleteClusterCommand_1 = require("./commands/DeleteClusterCommand");
const DeleteFargateProfileCommand_1 = require("./commands/DeleteFargateProfileCommand");
const DeleteNodegroupCommand_1 = require("./commands/DeleteNodegroupCommand");
const DeregisterClusterCommand_1 = require("./commands/DeregisterClusterCommand");
const DescribeAddonCommand_1 = require("./commands/DescribeAddonCommand");
const DescribeAddonConfigurationCommand_1 = require("./commands/DescribeAddonConfigurationCommand");
const DescribeAddonVersionsCommand_1 = require("./commands/DescribeAddonVersionsCommand");
const DescribeClusterCommand_1 = require("./commands/DescribeClusterCommand");
const DescribeFargateProfileCommand_1 = require("./commands/DescribeFargateProfileCommand");
const DescribeIdentityProviderConfigCommand_1 = require("./commands/DescribeIdentityProviderConfigCommand");
const DescribeNodegroupCommand_1 = require("./commands/DescribeNodegroupCommand");
const DescribeUpdateCommand_1 = require("./commands/DescribeUpdateCommand");
const DisassociateIdentityProviderConfigCommand_1 = require("./commands/DisassociateIdentityProviderConfigCommand");
const ListAddonsCommand_1 = require("./commands/ListAddonsCommand");
const ListClustersCommand_1 = require("./commands/ListClustersCommand");
const ListFargateProfilesCommand_1 = require("./commands/ListFargateProfilesCommand");
const ListIdentityProviderConfigsCommand_1 = require("./commands/ListIdentityProviderConfigsCommand");
const ListNodegroupsCommand_1 = require("./commands/ListNodegroupsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListUpdatesCommand_1 = require("./commands/ListUpdatesCommand");
const RegisterClusterCommand_1 = require("./commands/RegisterClusterCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateAddonCommand_1 = require("./commands/UpdateAddonCommand");
const UpdateClusterConfigCommand_1 = require("./commands/UpdateClusterConfigCommand");
const UpdateClusterVersionCommand_1 = require("./commands/UpdateClusterVersionCommand");
const UpdateNodegroupConfigCommand_1 = require("./commands/UpdateNodegroupConfigCommand");
const UpdateNodegroupVersionCommand_1 = require("./commands/UpdateNodegroupVersionCommand");
const EKSClient_1 = require("./EKSClient");
const commands = {
    AssociateEncryptionConfigCommand: AssociateEncryptionConfigCommand_1.AssociateEncryptionConfigCommand,
    AssociateIdentityProviderConfigCommand: AssociateIdentityProviderConfigCommand_1.AssociateIdentityProviderConfigCommand,
    CreateAddonCommand: CreateAddonCommand_1.CreateAddonCommand,
    CreateClusterCommand: CreateClusterCommand_1.CreateClusterCommand,
    CreateFargateProfileCommand: CreateFargateProfileCommand_1.CreateFargateProfileCommand,
    CreateNodegroupCommand: CreateNodegroupCommand_1.CreateNodegroupCommand,
    DeleteAddonCommand: DeleteAddonCommand_1.DeleteAddonCommand,
    DeleteClusterCommand: DeleteClusterCommand_1.DeleteClusterCommand,
    DeleteFargateProfileCommand: DeleteFargateProfileCommand_1.DeleteFargateProfileCommand,
    DeleteNodegroupCommand: DeleteNodegroupCommand_1.DeleteNodegroupCommand,
    DeregisterClusterCommand: DeregisterClusterCommand_1.DeregisterClusterCommand,
    DescribeAddonCommand: DescribeAddonCommand_1.DescribeAddonCommand,
    DescribeAddonConfigurationCommand: DescribeAddonConfigurationCommand_1.DescribeAddonConfigurationCommand,
    DescribeAddonVersionsCommand: DescribeAddonVersionsCommand_1.DescribeAddonVersionsCommand,
    DescribeClusterCommand: DescribeClusterCommand_1.DescribeClusterCommand,
    DescribeFargateProfileCommand: DescribeFargateProfileCommand_1.DescribeFargateProfileCommand,
    DescribeIdentityProviderConfigCommand: DescribeIdentityProviderConfigCommand_1.DescribeIdentityProviderConfigCommand,
    DescribeNodegroupCommand: DescribeNodegroupCommand_1.DescribeNodegroupCommand,
    DescribeUpdateCommand: DescribeUpdateCommand_1.DescribeUpdateCommand,
    DisassociateIdentityProviderConfigCommand: DisassociateIdentityProviderConfigCommand_1.DisassociateIdentityProviderConfigCommand,
    ListAddonsCommand: ListAddonsCommand_1.ListAddonsCommand,
    ListClustersCommand: ListClustersCommand_1.ListClustersCommand,
    ListFargateProfilesCommand: ListFargateProfilesCommand_1.ListFargateProfilesCommand,
    ListIdentityProviderConfigsCommand: ListIdentityProviderConfigsCommand_1.ListIdentityProviderConfigsCommand,
    ListNodegroupsCommand: ListNodegroupsCommand_1.ListNodegroupsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    ListUpdatesCommand: ListUpdatesCommand_1.ListUpdatesCommand,
    RegisterClusterCommand: RegisterClusterCommand_1.RegisterClusterCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateAddonCommand: UpdateAddonCommand_1.UpdateAddonCommand,
    UpdateClusterConfigCommand: UpdateClusterConfigCommand_1.UpdateClusterConfigCommand,
    UpdateClusterVersionCommand: UpdateClusterVersionCommand_1.UpdateClusterVersionCommand,
    UpdateNodegroupConfigCommand: UpdateNodegroupConfigCommand_1.UpdateNodegroupConfigCommand,
    UpdateNodegroupVersionCommand: UpdateNodegroupVersionCommand_1.UpdateNodegroupVersionCommand,
};
class EKS extends EKSClient_1.EKSClient {
}
exports.EKS = EKS;
(0, smithy_client_1.createAggregatedClient)(commands, EKS);
