// @aws-quickstart/eks-blueprints/dist/addons/ack/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AckAddOn = void 0;
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const ts_deepmerge_1 = require("ts-deepmerge");
require("reflect-metadata");
const utils_1 = require("../../utils");
const helm_addon_1 = require("../helm-addon");
const serviceMappings_1 = require("./serviceMappings");
__exportStar(require("./serviceMappings"), exports);
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    namespace: "ack-system",
    values: {},
    createNamespace: true,
    serviceName: serviceMappings_1.AckServiceName.IAM,
    id: "iam-ack"
};
/**
 * Main class to instantiate the Helm chart
 */
let AckAddOn = exports.AckAddOn = class AckAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super(populateDefaults(defaultProps, props));
        this.options = this.props;
        this.id = this.options.id;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        const sa = cluster.addServiceAccount(`${this.options.chart}-sa`, {
            namespace: this.options.namespace,
            name: this.options.saName,
        });
        let values = populateValues(this.options, cluster.stack.region);
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        if (this.options.createNamespace == true) {
            // Let CDK Create the Namespace
            const namespace = (0, utils_1.createNamespace)(this.options.namespace, cluster);
            sa.node.addDependency(namespace);
        }
        sa.role.addManagedPolicy(aws_iam_1.ManagedPolicy.fromAwsManagedPolicyName(this.options.managedPolicyName));
        const chart = this.addHelmChart(clusterInfo, values);
        chart.node.addDependency(sa);
        return Promise.resolve(chart);
    }
};
exports.AckAddOn = AckAddOn = __decorate([
    Reflect.metadata("ordered", true)
], AckAddOn);
/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions, awsRegion) {
    var _a;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, utils_1.setPath)(values, "aws.region", awsRegion);
    (0, utils_1.setPath)(values, "serviceAccount.create", false);
    (0, utils_1.setPath)(values, "serviceAccount.name", helmOptions.saName);
    return values;
}
/**
 * populate parameters passed or the default values from service Mappings.
 */
function populateDefaults(defaultProps, props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    let tempProps = { ...props !== null && props !== void 0 ? props : {} }; // since props may be empty
    tempProps.id = (_a = tempProps.id) !== null && _a !== void 0 ? _a : defaultProps.id;
    tempProps.serviceName = (_b = tempProps.serviceName) !== null && _b !== void 0 ? _b : defaultProps.serviceName;
    tempProps.name = (_c = tempProps.name) !== null && _c !== void 0 ? _c : serviceMappings_1.serviceMappings[tempProps.serviceName].chart;
    tempProps.namespace = (_d = tempProps.namespace) !== null && _d !== void 0 ? _d : defaultProps.namespace;
    tempProps.chart = (_e = tempProps.chart) !== null && _e !== void 0 ? _e : (_f = serviceMappings_1.serviceMappings[tempProps.serviceName]) === null || _f === void 0 ? void 0 : _f.chart;
    tempProps.version = (_g = tempProps.version) !== null && _g !== void 0 ? _g : (_h = serviceMappings_1.serviceMappings[tempProps.serviceName]) === null || _h === void 0 ? void 0 : _h.version;
    const repositoryUrl = "oci://public.ecr.aws/aws-controllers-k8s";
    tempProps.release = (_j = tempProps.release) !== null && _j !== void 0 ? _j : tempProps.chart;
    tempProps.repository = (_k = tempProps.repository) !== null && _k !== void 0 ? _k : `${repositoryUrl}/${tempProps.name}`;
    tempProps.managedPolicyName = (_l = tempProps.managedPolicyName) !== null && _l !== void 0 ? _l : (_m = serviceMappings_1.serviceMappings[tempProps.serviceName]) === null || _m === void 0 ? void 0 : _m.managedPolicyName;
    tempProps.createNamespace = (_o = tempProps.createNamespace) !== null && _o !== void 0 ? _o : defaultProps.createNamespace;
    tempProps.saName = (_p = tempProps.saName) !== null && _p !== void 0 ? _p : `${tempProps.chart}-sa`;
    return tempProps;
}
