// @aws-quickstart/eks-blueprints/dist/addons/external-secrets/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalsSecretsAddOn = void 0;
const utils_1 = require("../../utils");
const iam = require("aws-cdk-lib/aws-iam");
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
const aws_cdk_lib_1 = require("aws-cdk-lib");
/**
 * Default props for the add-on.
 */
const defaultProps = {
    name: "external-secrets",
    chart: "external-secrets",
    release: "blueprints-addon-external-secrets",
    version: "0.9.3",
    repository: "https://charts.external-secrets.io",
    namespace: "external-secrets",
    values: {},
};
/**
 * Default iam policy
 */
const defaultIamPolicy = new iam.PolicyStatement({
    effect: iam.Effect.ALLOW,
    actions: [
        "secretsmanager:GetResourcePolicy",
        "secretsmanager:GetSecretValue",
        "secretsmanager:DescribeSecret",
        "secretsmanager:ListSecretVersionIds",
        "secretsmanager:ListSecrets",
        "ssm:DescribeParameters",
        "ssm:GetParameter",
        "ssm:GetParameters",
        "ssm:GetParametersByPath",
        "ssm:GetParameterHistory",
        "kms:Decrypt"
    ],
    resources: ["*"],
});
/**
 * ExternalsSecretsAddOn deploys ExternalsSecrets into an EKS cluster using the `external-secrets` Helm chart.
 * https://github.com/external-secrets/external-secrets/
 *
 * For information on how to configure the `external-secrets` Helm chart, please view the values.yaml spec provided by the chart.
 * https://github.com/external-secrets/external-secrets/blob/main/deploy/charts/external-secrets/values.yaml
 */
class ExternalsSecretsAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        // Create the ExternalsSecrets namespace.
        const namespace = this.options.namespace;
        const ns = (0, utils_1.createNamespace)(this.options.namespace, cluster, true);
        // Create the ExternalsSecrets service account.
        const serviceAccountName = "external-secrets-sa";
        const sa = cluster.addServiceAccount(serviceAccountName, {
            name: serviceAccountName,
            namespace: namespace,
        });
        sa.node.addDependency(ns);
        // Apply additional IAM policies to the service account.
        const policies = this.options.iamPolicies || [defaultIamPolicy];
        policies.forEach((policy) => sa.addToPrincipalPolicy(policy));
        // Configure values.
        let values = {
            serviceAccount: {
                name: serviceAccountName,
                create: false,
            },
            ...this.options.values,
        };
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const helmChart = this.addHelmChart(clusterInfo, values, false, true, aws_cdk_lib_1.Duration.minutes(15));
        helmChart.node.addDependency(sa);
        return Promise.resolve(helmChart);
    }
}
exports.ExternalsSecretsAddOn = ExternalsSecretsAddOn;
