// @aws-quickstart/eks-blueprints/dist/addons/grafana-operator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrafanaOperatorAddon = void 0;
const ts_deepmerge_1 = require("ts-deepmerge");
const utils_1 = require("../../utils");
const helm_addon_1 = require("../helm-addon");
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    name: 'grafana-operator',
    chart: 'oci://ghcr.io/grafana-operator/helm-charts/grafana-operator',
    namespace: 'grafana-operator',
    release: 'grafana-operator',
    version: 'v5.0.0-rc3',
    values: {},
    createNamespace: true
};
/**
 * Main class to instantiate the Helm chart
 */
class GrafanaOperatorAddon extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a, _b;
        const cluster = clusterInfo.cluster;
        let values = (_a = this.options.values) !== null && _a !== void 0 ? _a : {};
        values = (0, ts_deepmerge_1.default)(values, (_b = this.props.values) !== null && _b !== void 0 ? _b : {});
        const chart = this.addHelmChart(clusterInfo, values);
        if (this.options.createNamespace == true) {
            // Let CDK Create the Namespace
            const namespace = (0, utils_1.createNamespace)(this.options.namespace, cluster);
            chart.node.addDependency(namespace);
        }
        return Promise.resolve(chart);
    }
}
exports.GrafanaOperatorAddon = GrafanaOperatorAddon;
