// @aws-quickstart/eks-blueprints/dist/utils/proxy-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveTarget = exports.DummyProxy = exports.sourceFunction = exports.isDynamicProxy = void 0;
const nutil = require("node:util/types");
/**
 * Symbol that uniquely designates that a particular proxy is instance of our DummyProxy
 */
exports.isDynamicProxy = Symbol("isDynamicProxy");
/**
 * Symbol that retrieves the source function from the proxy. This function is expected to create the required target (e.g. resource).
 */
exports.sourceFunction = Symbol("sourceFunction");
/**
 * Simple proxy implementation that will require resolution at runtime (enables lazy loading).
 * Unlike dynamic proxy that can create target on the fly, this proxy
 * just a place-holder that supplies the function that can be used to resolve the target.
 * Since most CDK constructs are not idempotent (meaning you can not call a create function twice, the second will fail)
 * this design choice was the simplest to support declarative resources.
 * Customers can clone the supplied JSON structure with cloneDeep and replace proxies with the actual targets as part of that process.
 */
class DummyProxy {
    constructor(source) {
        this.source = source;
    }
    get(_, key) {
        if (key === exports.isDynamicProxy) {
            return true;
        }
        if (key === exports.sourceFunction) {
            return this.source;
        }
        return new Proxy({}, new DummyProxy((arg) => {
            return this.source(arg)[key];
        }));
    }
}
exports.DummyProxy = DummyProxy;
/**
 * Function resolves the proxy with the target, that enables lazy loading use cases.
 * @param value potential proxy to resolve
 * @param arg represents the argument that should be passed to the resolution function (sourceFunction).
 * @returns
 */
function resolveTarget(value, arg) {
    if (nutil.isProxy(value)) {
        const object = value;
        if (object[exports.isDynamicProxy]) {
            const fn = object[exports.sourceFunction];
            return fn(arg);
        }
    }
    return value;
}
exports.resolveTarget = resolveTarget;
