// @aws-quickstart/eks-blueprints/dist/addons/knative-operator/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KNativeOperator = void 0;
const utils_1 = require("../../utils");
const kubectl_provider_1 = require("../helm-addon/kubectl-provider");
const defaultProps = {
    name: 'knative-operator',
    namespace: 'default',
    version: 'v1.8.1',
};
/**
 * Implementation of KNative add-on for EKS Blueprints. Installs KNative to the Cluster.
 */
class KNativeOperator {
    constructor(props) {
        this.knativeAddOnProps = { ...defaultProps, ...props };
    }
    deploy(clusterInfo) {
        const BASE_URL = `https://github.com/knative/operator/releases/download/knative`;
        // Load External YAML: https://github.com/knative/operator/releases/download/knative-v1.8.1/operator.yaml
        const doc = (0, utils_1.loadExternalYaml)(BASE_URL + `-${this.knativeAddOnProps.version}/operator.yaml`).slice(0, 26); // the last element is null
        const kubectlProvider = new kubectl_provider_1.KubectlProvider(clusterInfo);
        const statement = kubectlProvider.addManifest({ manifest: doc, values: {}, name: 'knative-operator', namespace: this.knativeAddOnProps.namespace });
        return Promise.resolve(statement);
    }
}
exports.KNativeOperator = KNativeOperator;
__decorate([
    (0, utils_1.dependable)('IstioControlPlaneAddOn')
], KNativeOperator.prototype, "deploy", null);
