// @aws-quickstart/eks-blueprints/dist/addons/secrets-store/csi-driver-provider-aws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CsiDriverProviderAws = void 0;
const cdk = require("aws-cdk-lib");
const yaml_utils_1 = require("../../utils/yaml-utils");
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
class CsiDriverProviderAws {
    constructor(props) {
        this.props = props;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        let values = {
            grpcSupportedProviders: 'aws'
        };
        if (typeof (this.props.rotationPollInterval) === 'string') {
            values.enableSecretRotation = 'true';
            values.rotationPollInterval = this.props.rotationPollInterval;
        }
        if (this.props.syncSecrets === true) {
            values.syncSecret = {
                enabled: 'true'
            };
        }
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const helmChartOptions = {
            chart: this.props.chart,
            repository: this.props.repository,
            namespace: this.props.namespace,
            release: this.props.release,
            version: this.props.version,
            wait: true,
            timeout: cdk.Duration.minutes(15),
            values,
        };
        helm_addon_1.HelmAddOn.validateVersion({
            chart: helmChartOptions.chart,
            repository: helmChartOptions.repository,
            version: helmChartOptions.version
        });
        const secretStoreCSIDriverHelmChart = cluster.addHelmChart('SecretsStoreCSIDriver', helmChartOptions);
        const manifestUrl = this.props.ascpUrl;
        const manifest = (0, yaml_utils_1.loadExternalYaml)(manifestUrl);
        const secretProviderManifest = clusterInfo.cluster.addManifest('SecretsStoreCsiDriverProviderAws', ...manifest);
        secretProviderManifest.node.addDependency(secretStoreCSIDriverHelmChart);
        return secretProviderManifest;
    }
}
exports.CsiDriverProviderAws = CsiDriverProviderAws;
