// @aws-quickstart/eks-blueprints/dist/addons/metrics-server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricsServerAddOn = void 0;
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
const utils_1 = require("../../utils");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    chart: 'metrics-server',
    repository: 'https://kubernetes-sigs.github.io/metrics-server',
    version: '3.11.0',
    release: 'blueprints-addon-metrics-server',
    name: 'metrics-server',
    namespace: 'kube-system',
    createNamespace: false,
};
class MetricsServerAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a, _b;
        const cluster = clusterInfo.cluster;
        let values = (_a = this.options) !== null && _a !== void 0 ? _a : {};
        values = (0, ts_deepmerge_1.default)(values, (_b = this.props.values) !== null && _b !== void 0 ? _b : {});
        const chart = this.addHelmChart(clusterInfo, values);
        if (this.options.createNamespace == true) {
            // Let CDK Create the Namespace
            const namespace = (0, utils_1.createNamespace)(this.options.namespace, cluster);
            chart.node.addDependency(namespace);
        }
        return Promise.resolve(chart);
    }
}
exports.MetricsServerAddOn = MetricsServerAddOn;
