// @aws-quickstart/eks-blueprints/dist/builders/observability-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObservabilityBuilder = void 0;
const stacks_1 = require("../stacks");
const addons = require("../addons");
const utils = require("../utils");
const aws_cdk_lib_1 = require("aws-cdk-lib");
class ObservabilityBuilder extends stacks_1.BlueprintBuilder {
    /**
     * This method helps you prepare a blueprint for setting up observability
     * returning an array of blueprint addons for AWS native services
     */
    enableNativePatternAddOns() {
        return this.addOns(new addons.AwsLoadBalancerControllerAddOn(), new addons.CertManagerAddOn(), new addons.ContainerInsightsAddOn(), new addons.CoreDnsAddOn(), new addons.KubeProxyAddOn(), new addons.KubeStateMetricsAddOn(), new addons.MetricsServerAddOn(), new addons.PrometheusNodeExporterAddOn());
    }
    /**
     * This method helps you prepare a blueprint for setting up observability
     * returning an array of blueprint addons for combination of AWS native and
     * AWS managed open source services
     */
    enableMixedPatternAddOns() {
        return this.addOns(new addons.AwsLoadBalancerControllerAddOn(), new addons.CertManagerAddOn(), new addons.AdotCollectorAddOn(), new addons.CoreDnsAddOn(), new addons.KubeProxyAddOn(), new addons.KubeStateMetricsAddOn(), new addons.MetricsServerAddOn(), new addons.PrometheusNodeExporterAddOn());
    }
    /**
     * This method helps you prepare a blueprint for setting up observability
     * returning an array of blueprint addons for AWS managed open source services
     * @param ampAddOnProps AmpAddonProps. This is mandatory.
     * @param kubeProxyVersion KubeProxy Version. This is optional.
     */
    enableOpenSourcePatternAddOns(ampAddOnProps, kubeProxyVersion) {
        return this.addOns(new addons.AwsLoadBalancerControllerAddOn(), new addons.CertManagerAddOn(), new addons.AdotCollectorAddOn(), new addons.AmpAddOn(ampAddOnProps), new addons.CoreDnsAddOn(), new addons.ExternalsSecretsAddOn(), new addons.GrafanaOperatorAddon(), new addons.KubeProxyAddOn(kubeProxyVersion), new addons.KubeStateMetricsAddOn(), new addons.MetricsServerAddOn(), new addons.PrometheusNodeExporterAddOn());
    }
    /**
     * This method helps you prepare a blueprint for setting up observability with
     * usage tracking addon
     */
    static builder() {
        const builder = new ObservabilityBuilder();
        builder.addOns(new addons.NestedStackAddOn({
            id: "usage-tracking-addon",
            builder: UsageTrackingAddOn.builder(),
        }));
        return builder;
    }
}
exports.ObservabilityBuilder = ObservabilityBuilder;
/**
 * Nested stack that is used as tracker for Observability Accelerator
 */
class UsageTrackingAddOn extends aws_cdk_lib_1.NestedStack {
    static builder() {
        return {
            build(scope, id, props) {
                return new UsageTrackingAddOn(scope, id, props);
            }
        };
    }
    constructor(scope, id, props) {
        super(scope, id, utils.withUsageTracking(UsageTrackingAddOn.USAGE_ID, props));
    }
}
UsageTrackingAddOn.USAGE_ID = "qs-1u9l12gj7";
