// @aws-quickstart/eks-blueprints/dist/addons/cert-manager/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertManagerAddOn = void 0;
const ts_deepmerge_1 = require("ts-deepmerge");
const utils_1 = require("../../utils");
const object_utils_1 = require("../../utils/object-utils");
const helm_addon_1 = require("../helm-addon");
const aws_loadbalancer_controller_1 = require("../aws-loadbalancer-controller");
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    name: "blueprints-cert-manager-addon",
    namespace: "cert-manager",
    chart: "cert-manager",
    version: "1.12.3",
    release: "cert-manager",
    repository: "https://charts.jetstack.io",
    values: {},
    installCRDs: true,
    createNamespace: true
};
/**
 * Main class to instantiate the Helm chart
 */
class CertManagerAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        let values = populateValues(this.options);
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        if (this.options.createNamespace == true) {
            // Let CDK Create the Namespace
            const namespace = (0, utils_1.createNamespace)(this.options.namespace, cluster);
            const chart = this.addHelmChart(clusterInfo, values);
            chart.node.addDependency(namespace);
            return Promise.resolve(chart);
        }
        else {
            //Namespace is already created
            const chart = this.addHelmChart(clusterInfo, values);
            return Promise.resolve(chart);
        }
    }
}
exports.CertManagerAddOn = CertManagerAddOn;
__decorate([
    (0, utils_1.dependable)(aws_loadbalancer_controller_1.AwsLoadBalancerControllerAddOn.name)
], CertManagerAddOn.prototype, "deploy", null);
/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions) {
    var _a;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, object_utils_1.setPath)(values, "installCRDs", helmOptions.installCRDs);
    return values;
}
