// @aws-quickstart/eks-blueprints/dist/addons/cloudwatch-adot-addon/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudWatchAdotAddOn = void 0;
const utils_1 = require("../../utils");
const adot_1 = require("../adot");
const kubectl_provider_1 = require("../helm-addon/kubectl-provider");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    deploymentMode: "deployment" /* cloudWatchDeploymentMode.DEPLOYMENT */,
    namespace: 'default',
    name: 'adot-collector-cloudwatch',
    metricsNameSelectors: ['apiserver_request_.*', 'container_memory_.*', 'container_threads', 'otelcol_process_.*'],
    podLabelRegex: '.*'
};
/**
 * Implementation of CloudWatch ADOT add-on for EKS Blueprints. Installs ADOT Collector.
 */
class CloudWatchAdotAddOn {
    constructor(props) {
        this.cloudWatchAddOnProps = { ...defaultProps, ...props };
    }
    deploy(clusterInfo) {
        var _a, _b;
        const cluster = clusterInfo.cluster;
        // Applying manifest for configuring ADOT Collector for CloudWatch.
        const doc = (0, utils_1.readYamlDocument)(__dirname + '/collector-config-cloudwatch.ytpl');
        const manifest = doc.split("---").map(e => (0, utils_1.loadYaml)(e));
        const values = {
            awsRegion: cluster.stack.region,
            deploymentMode: this.cloudWatchAddOnProps.deploymentMode,
            namespace: this.cloudWatchAddOnProps.namespace,
            clusterName: cluster.clusterName,
            metricsNameSelectors: (_a = this.cloudWatchAddOnProps.metricsNameSelectors) !== null && _a !== void 0 ? _a : defaultProps.metricsNameSelectors,
            podLabelRegex: (_b = this.cloudWatchAddOnProps.podLabelRegex) !== null && _b !== void 0 ? _b : defaultProps.podLabelRegex
        };
        const manifestDeployment = {
            name: this.cloudWatchAddOnProps.name,
            namespace: this.cloudWatchAddOnProps.namespace,
            manifest,
            values
        };
        const kubectlProvider = new kubectl_provider_1.KubectlProvider(clusterInfo);
        const statement = kubectlProvider.addManifest(manifestDeployment);
        return Promise.resolve(statement);
    }
}
exports.CloudWatchAdotAddOn = CloudWatchAdotAddOn;
__decorate([
    (0, utils_1.dependable)(adot_1.AdotCollectorAddOn.name)
], CloudWatchAdotAddOn.prototype, "deploy", null);
