// @aws-quickstart/eks-blueprints/dist/addons/container-insights/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerInsightsAddOn = void 0;
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const ts_deepmerge_1 = require("ts-deepmerge");
const helm_addon_1 = require("../helm-addon");
const utils_1 = require("../../utils");
const defaultProps = {
    name: "adot-exporter-for-eks-on-ec2",
    namespace: undefined,
    chart: "adot-exporter-for-eks-on-ec2",
    version: "0.15.0",
    release: "adot-eks-addon",
    repository: "https://aws-observability.github.io/aws-otel-helm-charts"
};
class ContainerInsightsAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
    }
    /**
     * @override
     */
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        const policy = aws_iam_1.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy');
        // Create an adot-collector service account.
        const serviceAccountName = "adot-collector-sa";
        let serviceAccountNamespace;
        if (this.props.namespace) {
            serviceAccountNamespace = this.props.namespace;
        }
        else {
            serviceAccountNamespace = "amazon-metrics";
        }
        const ns = (0, utils_1.createNamespace)(serviceAccountNamespace, cluster, true);
        const sa = cluster.addServiceAccount(serviceAccountName, {
            name: serviceAccountName,
            namespace: serviceAccountNamespace,
        });
        // Apply Managed IAM policy to the service account.
        sa.role.addManagedPolicy(policy);
        sa.node.addDependency(ns);
        let values = {
            awsRegion: cluster.stack.region,
            clusterName: cluster.clusterName,
            serviceAccount: {
                create: false,
            },
            adotCollector: {
                daemonSet: {
                    createNamespace: false,
                    service: {
                        metrics: {
                            receivers: ["awscontainerinsightreceiver"],
                            exporters: ["awsemf"],
                        }
                    },
                    serviceAccount: {
                        create: false,
                    },
                    cwexporters: {
                        logStreamName: "EKSNode",
                    }
                }
            }
        };
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const chart = this.addHelmChart(clusterInfo, values, true, false);
        chart.node.addDependency(sa);
        return Promise.resolve(chart);
    }
}
exports.ContainerInsightsAddOn = ContainerInsightsAddOn;
__decorate([
    (0, utils_1.conflictsWith)("AdotCollectorAddOn")
], ContainerInsightsAddOn.prototype, "deploy", null);
