// @aws-quickstart/eks-blueprints/dist/addons/istio-base/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IstioBaseAddOn = void 0;
const helm_addon_1 = require("../helm-addon");
const namespace_utils_1 = require("../../utils/namespace-utils");
const ts_deepmerge_1 = require("ts-deepmerge");
const aws_cdk_lib_1 = require("aws-cdk-lib");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    name: "istio-base",
    release: "istio-base",
    namespace: "istio-system",
    chart: "base",
    version: "1.18.2",
    repository: "https://istio-release.storage.googleapis.com/charts"
};
class IstioBaseAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        // Istio Namespace
        const namespace = (0, namespace_utils_1.createNamespace)('istio-system', cluster);
        let values = {
            global: {
                istiod: {
                    enableAnalysis: this.options.enableAnalysis
                },
                configValidation: this.options.configValidation,
                externalIstiod: this.options.externalIstiod,
                base: {
                    enableIstioConfigCRDs: this.options.enableIstioConfigCRDs
                }
            }
        };
        values = (0, ts_deepmerge_1.default)(values, (_a = this.props.values) !== null && _a !== void 0 ? _a : {});
        const chart = this.addHelmChart(clusterInfo, values, undefined, true, aws_cdk_lib_1.Duration.seconds(60));
        chart.node.addDependency(namespace);
        return Promise.resolve(chart);
    }
}
exports.IstioBaseAddOn = IstioBaseAddOn;
