// @aws-quickstart/eks-blueprints/dist/addons/fluxcd/gitrepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluxGitRepository = void 0;
const utils_1 = require("../../utils");
/**
 * Flux GitRepository API defines a Source to produce an Artifact for a Git repository revision.
 */
class FluxGitRepository {
    constructor(repository) {
        this.repository = repository;
    }
    generate(name, namespace, fluxSyncInterval, fluxSecretRefName) {
        const repository = this.repository;
        const gitManifest = {
            apiVersion: "source.toolkit.fluxcd.io/v1beta2",
            kind: "GitRepository",
            metadata: {
                name: name,
                namespace: namespace
            },
            spec: {
                interval: fluxSyncInterval,
                url: repository.repoUrl,
                ref: {
                    branch: repository.targetRevision,
                },
            }
        };
        if (fluxSecretRefName) {
            (0, utils_1.setPath)(gitManifest, "spec.secretRef.name", fluxSecretRefName);
        }
        return gitManifest;
    }
}
exports.FluxGitRepository = FluxGitRepository;
