// @aws-quickstart/eks-blueprints/dist/addons/fluxcd/kustomization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluxKustomization = void 0;
const utils_1 = require("../../utils");
/**
 * Flux Kustomization API defines a pipeline for fetching, decrypting, building, validating and applying Kustomize overlays or plain Kubernetes manifests.
 */
class FluxKustomization {
    constructor() { }
    generate(name, repoName, namespace, fluxSyncInterval, fluxPrune, fluxTimeout, values, fluxKustomizationPath, fluxTargetNamespace) {
        const kustomizationManifest = {
            apiVersion: "kustomize.toolkit.fluxcd.io/v1beta2",
            kind: "Kustomization",
            metadata: {
                name,
                namespace
            },
            spec: {
                interval: fluxSyncInterval,
                sourceRef: {
                    kind: "GitRepository",
                    name: repoName
                },
                path: fluxKustomizationPath,
                prune: fluxPrune,
                timeout: fluxTimeout
            }
        };
        if (values) {
            (0, utils_1.setPath)(kustomizationManifest, "spec.postBuild.substitute", values);
        }
        if (fluxTargetNamespace) {
            (0, utils_1.setPath)(kustomizationManifest, "spec.targetNamespace", fluxTargetNamespace);
        }
        return kustomizationManifest;
    }
}
exports.FluxKustomization = FluxKustomization;
