// @aws-quickstart/eks-blueprints/dist/addons/ebs-csi-driver/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EbsCsiDriverAddOn = void 0;
const core_addon_1 = require("../core-addon");
const iam_policy_1 = require("./iam-policy");
/**
 * Default values for the add-on
 */
const defaultProps = {
    addOnName: "aws-ebs-csi-driver",
    version: "v1.21.0-eksbuild.1",
    saName: "ebs-csi-controller-sa",
};
/**
 * Implementation of EBS CSI Driver EKS add-on
 */
class EbsCsiDriverAddOn extends core_addon_1.CoreAddOn {
    constructor(options) {
        var _a;
        super({
            addOnName: defaultProps.addOnName,
            version: (_a = options === null || options === void 0 ? void 0 : options.version) !== null && _a !== void 0 ? _a : defaultProps.version,
            saName: defaultProps.saName
        });
        this.options = options;
    }
    providePolicyDocument(clusterInfo) {
        var _a;
        return (0, iam_policy_1.getEbsDriverPolicyDocument)(clusterInfo.cluster.stack.partition, (_a = this.options) === null || _a === void 0 ? void 0 : _a.kmsKeys);
    }
}
exports.EbsCsiDriverAddOn = EbsCsiDriverAddOn;
