// @aws-quickstart/eks-blueprints/dist/resource-providers/efs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LookupEfsFileSystemProvider = exports.CreateEfsFileSystemProvider = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const ec2 = require("aws-cdk-lib/aws-ec2");
const efs = require("aws-cdk-lib/aws-efs");
const spi_1 = require("../spi");
/**
 * EFS resource provider.
 *
 * @param name The name of the EFS file system to create.
 * @param efsProps The props used for the file system.
 */
class CreateEfsFileSystemProvider {
    constructor(options) {
        this.options = options;
    }
    provide(context) {
        var _a;
        const id = context.scope.node.id;
        const securityGroupId = `${id}-${(_a = this.options.name) !== null && _a !== void 0 ? _a : "default"}-EfsSecurityGroup`;
        let efsFileSystem;
        const vpc = context.get(spi_1.GlobalResources.Vpc);
        if (vpc === undefined) {
            throw new Error("VPC not found in context");
        }
        const clusterVpcCidr = vpc.vpcCidrBlock;
        let kmsKey;
        if (this.options.kmsKeyResourceName) {
            kmsKey = context.get(this.options.kmsKeyResourceName);
        }
        const efsSG = new ec2.SecurityGroup(context.scope, securityGroupId, {
            vpc,
            securityGroupName: securityGroupId,
        });
        efsSG.addIngressRule(ec2.Peer.ipv4(clusterVpcCidr), new ec2.Port({
            protocol: ec2.Protocol.TCP,
            stringRepresentation: "EFSconnection",
            toPort: 2049,
            fromPort: 2049,
        }));
        efsFileSystem = new efs.FileSystem(context.scope, this.options.name || `${id}-EfsFileSystem`, {
            vpc,
            securityGroup: efsSG,
            kmsKey,
            ...this.options.efsProps,
        });
        new aws_cdk_lib_1.CfnOutput(context.scope, "EfsFileSystemId", {
            value: efsFileSystem.fileSystemId,
        });
        return efsFileSystem;
    }
}
exports.CreateEfsFileSystemProvider = CreateEfsFileSystemProvider;
/**
 * Pass an EFS file system name and id to lookup an existing EFS file system.
 * @param name The name of the EFS file system to lookup an existing EFS file system.
 * @param fileSystemId The id of the EFS file system to lookup an existing EFS file system.
 */
class LookupEfsFileSystemProvider {
    constructor(options) {
        this.options = options;
    }
    provide(context) {
        var _a;
        const id = context.scope.node.id;
        const securityGroupId = `${id}-${(_a = this.options.name) !== null && _a !== void 0 ? _a : "default"}-EfsSecurityGroup`;
        let efsFileSystem;
        const securityGroup = ec2.SecurityGroup.fromSecurityGroupId(context.scope, securityGroupId, securityGroupId);
        efsFileSystem = efs.FileSystem.fromFileSystemAttributes(context.scope, this.options.name, {
            securityGroup: securityGroup,
            fileSystemId: this.options.fileSystemId,
        });
        new aws_cdk_lib_1.CfnOutput(context.scope, "EfsFileSystemId", {
            value: efsFileSystem.fileSystemId,
        });
        return efsFileSystem;
    }
}
exports.LookupEfsFileSystemProvider = LookupEfsFileSystemProvider;
