// @aws-quickstart/eks-blueprints/dist/resource-providers/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResource = exports.getNamedResource = void 0;
const proxy_utils_1 = require("../utils/proxy-utils");
const uuid_1 = require("uuid");
/**
 * Creates a proxy to the named resource provider. This proxy will resolve to the type of the
 * resource provider under the provided name.
 * It enables getting references to resources outside of the Stack construct and using them with the blueprint:
 * @example
 * const app = new cdk.App();
 * const adminRole: iam.IRole = blueprints.getNamedResource("my-admin-role""); // note, there is no stack class here
 *
 * const clusterProvider = new blueprints.GenericClusterProvider({
     mastersRole: adminRole,
     ... other props
 * });
 * @param resourceName
 * @returns
 */
function getNamedResource(resourceName) {
    return new Proxy({}, new proxy_utils_1.DummyProxy((resourceContext) => {
        return resourceContext.get(resourceName);
    }));
}
exports.getNamedResource = getNamedResource;
/**
 * Creates a proxy to an anonymous resource. This function allows passing the provider function as input.
 * It enables creating ad-hoc references outside of the Stack construct and using them with the blueprint.
 * Designed for cases when resource is defined once and needed in a single place.
 * @example
 * const app = new cdk.App();
 * const clusterProvider = new blueprints.GenericClusterProvider({
 *   mastersRole: blueprints.getResource(context => { // will generate a unique name for resource.
        return new iam.Role(context.scope, 'AdminRole', { assumedBy: new AccountRootPrincipal()});
    }),
    ... other props
});
 * @param resourceName
 * @returns
 */
function getResource(fn) {
    const uid = (0, uuid_1.v4)();
    return new Proxy({}, new proxy_utils_1.DummyProxy((resourceContext) => {
        let result = resourceContext.get(uid);
        if (result == null) {
            resourceContext.add(uid, {
                provide(context) {
                    return fn(context);
                }
            });
            result = resourceContext.get(uid);
        }
        return result;
    }));
}
exports.getResource = getResource;
